Support S3 namespaces when retrieving models from buckets

cianclarke opened the pull request:
**Proposed changes**:
- When working with many models, it helps to make use of S3 namespacing functionality (aka "folders") within buckets. Not being able to load models from buckets is a hindrance when working across projects. 
- Present S3 bucket support only works with root directory - whenever a namespace is provided, this path is also used on the local file system to write the `.tar.gz` (see #4916 )
- PR allows passing of S3 paths with namespaces for modal retrieval. 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

Fixes #4916

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4917) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4917) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=4917) it.</sub>

erohmensing commented:
Thanks for the PR, we'll give it a review as soon as possible.

ricwo commented on rasa/nlu/persistor.py:
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
 
-    def _retrieve_tar(self, target_filename: Text) -> None:
+    def _retrieve_tar(self, model_full_path: Text) -> None:
+        tar_name = os.path.basename(model_full_path)
```
this line should move one down (below the docstring)

cianclarke replied:
Done!

ricwo commented on tests/nlu/base/test_persistor.py:
```diff
@@ -142,3 +142,27 @@ def test_retrieve_tar_archive(model, archive):
+def test_s3_private_retrieve_tar():
+    model = "/my/s3/project/model.tar.gz"
+    awsPersistor = persistor.AWSPersistor("rasa-test")
+    with patch.object(awsPersistor.bucket, "download_fileobj") as download_fileobj:
```
```suggestion
    with patch.object(awsPersistor.bucket, "download_fileobj") as download_fileobj:
        # noinspection PyProtectedMember
```

ricwo commented on tests/nlu/base/test_persistor.py:
```diff
@@ -142,3 +142,27 @@ def test_retrieve_tar_archive(model, archive):
+        awsPersistor._retrieve_tar(model)
+    retrieveArgs = download_fileobj.call_args[0]
+    assert retrieveArgs[0] == model
+    assert retrieveArgs[1].name == "model.tar.gz"
```
can we move these two tests right below the existing two s3 tests?

cianclarke replied:
Done!

ricwo commented on tests/nlu/base/test_persistor.py:
```diff
@@ -142,3 +142,27 @@ def test_retrieve_tar_archive(model, archive):
+
+# noinspection PyPep8Naming
+@mock_s3
+def test_s3_private_retrieve_tar():
```
what exactly is being tested here? A comment would be very helpful

cianclarke replied:
Added! I included this test retroactively to bump coverage back to an acceptable level by coveralls. I'd be happy to also remove it if you think it excessive

cianclarke commented:
@ricwo Many thanks for the review. I've incorporated all your suggestions, and also added a changelog entry 👍

ricwo commented on rasa/nlu/persistor.py:
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
             self.s3.Object(self.bucket_name, file_key).put(Body=f)
 
-    def _retrieve_tar(self, target_filename: Text) -> None:
+    def _retrieve_tar(self, model_full_path: Text) -> None:
```
```suggestion
    def _retrieve_tar(self, model_path: Text) -> None:
```

ricwo replied:
in line with `model_path` args used elsewhere

ricwo replied:
@cianclarke would you mind updating the args?

ricwo commented on rasa/nlu/persistor.py:
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
-
-        with open(target_filename, "wb") as f:
-            self.bucket.download_fileobj(target_filename, f)
+        tar_name = os.path.basename(model_full_path)
```
```suggestion
        tar_name = os.path.basename(model_path)
```

ricwo commented on rasa/nlu/persistor.py:
```diff
@@ -151,11 +151,11 @@ def _persist_tar(self, file_key: Text, tar_path: Text) -> None:
-            self.bucket.download_fileobj(target_filename, f)
+        tar_name = os.path.basename(model_full_path)
+        with open(tar_name, "wb") as f:
+            self.bucket.download_fileobj(model_full_path, f)
```
```suggestion
            self.bucket.download_fileobj(model_path, f)
```

ricwo commented:
@cianclarke looks like you didn't check in the changelog file - would you mind adding that?

cianclarke commented:
Ah, sorry @ricwo - you're absolutely right. I've committed the changelog entry. 
I'm also noticing coveralls oddly report coverage having gone back down again, just from moving the tests around when I addressed your PR feedback. I'm not sure what's going on there!

cianclarke commented:
@ricwo - I committed your args suggestions from the GitHub UI (looks like I may have missed those). 
Hoping that was the argument updates you had meant - happy to also rebase the excess commits out, or can do on merge - not sure if the project has a preference on this.

cianclarke commented:
Thanks @ricwo
